whitepapers/0011_accountant.md

```diff
@@ -0,0 +1,190 @@
+- If the transfer does not yet have a quorum of signatures, return a `Pending` status and mark the transfer as having a signature from the guardian.
+- If a transfer with the same `(emitter_chain, emitter_address, sequence)` tuple has already been committed, then check that the digest of the transfer matches the digest of the committed transfer.  If they match return a `Committed` status; otherwise return an `Error` status and emit an `Error` event.
+- If the transfer has a quorum of signatures, parse the tokenbridge payload and commit the transfer.
+- If committing the transfer fails (for example, if one of the accounts does not have a sufficient balance), return an `Error` status and emit an `Error` event.  Otherwise return a `Committed` status and emit a `Committed` event.
+
+If the guardian receives a `Committed` status for a transfer then it can immediately proceed to signing the VAA.  If it receives an `Error` status, then it should halt further processing of the transfer (see the [Handling Rejections](#handling-rejections) section for more details).  If it receives a `Pending` status, then it should add the observation to a local database of pending transfers and wait for an event from the accountant.
+
+#### Observing Events from Wormhole Chain
+
+Each guardian sets up a watcher for the accountant watching for events emitted signalling a determination on a transfer.  An event for a transfer is emitted when there is a quorum of pre-observations.
+
+There are 2 types of events that the guardian must handle:
+
+- Transfer committed - This indicates that an observation has reached quorum and the transfer has been committed.  The guardian must verify the contents of the event and then sign the VAA if they match.  See [Threat Model](#threat-model) for more details.
+- Error - This indicates that the accountant ran into an error while processing an observation.  The guardian should halt further processing of the transfer.  See the [Handling Rejections](#handling-rejections) section for more details
+
+#### Handling stuck transfers
+
+It may be possible that some transfers don't receive enough observations to reach quorum.  This can happen if a number of guardians miss the transaction on the origin chain.  Since the accountant runs before the peer to peer gossip layer, it cannot take advantage of the existing automatic retry mechanisms that the guardian network uses.  To deal with this the accountant contract provides a `MissingObservations` query that returns a list of pending transfers for which the accountant does not have an observation from a particular guardian.
+
+Guardians are expected to periodically run this query and re-observe transactions for which the accountant is missing observations from them.
+
+#### Contract Upgrades
+
+Upgrading the contract will be done via a governance action from the guardian network with the same mechanism we use for the existing wormhole cosmwasm contracts.
```

Comment:
This part is no longer accurate.  The existing wormhole cosmwasm contracts do self-upgrades via governance VAA whereas wormchain contracts use x/wormhole [`MigrateContract`](https://github.com/wormhole-foundation/wormhole/blob/a846036b6ebff3af6f12ff375f5c3801ada20291/wormchain/x/wormhole/client/cli/tx_wasmd.go#L148) txs.

Reply:
So poking at the wormchain and guardian node code, my understanding of the upgrade process is:

Someone creates a MigrateContract governance vaa and distributes the prototxt.

All guardians sign the migrate contract vaa:
```bash
guardiand admin governance-vaa-inject migratecontract.prototxt
```

When 13 guardians have signed that VAA via the aforementioned step, someone with an allowed wallet can run [MigrateContract](https://github.com/wormhole-foundation/wormhole/blob/a846036b6ebff3af6f12ff375f5c3801ada20291/wormchain/x/wormhole/client/cli/tx_wasmd.go#L148) with the VAA and the wasm binary for the new contract. I'll try to update the text to represent the new process.

Reply:
Yes, that is my understanding as well but @conorpp would know for sure.

Reply:
I was a bit more verbose, but think I captured the gist of this in 5133abeb5ed1a3c648e5721b0bc69628c1d15abe so I'm going to resolve this. Let me know if it needs further clarification.

Reply:
yes this is correct.  The mechanism for supplying the governance & update is a bit different (e.g. accountant uses custom x/wormhole tx to submit).  I don't think it's too important to go into this detail but ok if we do.